### PR TITLE
[*] Replaces find_matching_token implementation with more performant version

### PR DESCRIPTION
**Pull request by PhilipSkinner**
### Summary

Performance of find_matching_token a major issue - can take systems offline.

This patch rewrites it to only fetch 1 valid access token from the database.

### Other Information

I used the existing tests as coverage over this particular modification - all tests are still passing.

I did notice that in order to get the tests to pass you must accept that an application may have no scopes, and if it has no scopes then no verification of the incoming scopes is carried out, nor does the system verify that the token being returned contains only claims available to the application.

Let me know if this is an oversight of the current system and I will correct this behavior - right now it works as it did before.

### CODE REVIEW

**Comment by PhilipSkinner** [comment]
Apologies, I wrote this with a feature that isn't readily available on older versions of active record, I've corrected it to work with older versions.

**Comment by PhilipSkinner** [comment]
Apologies for all the amends, ruby is a new language for me.

Let me know if there are any questions or modifications that I need to make.

File: lib/doorkeeper/models/access_token_mixin.rb [hunk 1]
```diff
@@ -120,44 +106,113 @@ def find_access_token_in_batches(relation, **args, &block)
...
+      #   set of scopes
+      #
+      # @return [ActiveRecord::Relation] filtered relation
+      #   query
+      #
+      def filter_by_application_scopes(relation, application, scopes)
+        application.scopes.each do |scope_val|
+          next unless scopes.exists? scope_val
+
+          relation = relation.where([
```
Comment (nbulaj) [inline lib/doorkeeper/models/access_token_mixin.rb:158]: I don't sure actually that this implementation matches previous one with `Doorkeeper::OAuth::Helpers::ScopeChecker
` :thinking: Even with passing specs it looks scary.. I have to to a deep dive into the changes to understand if everything is OK

Comment (PhilipSkinner) [reply lib/doorkeeper/models/access_token_mixin.rb:158]: Ok, I can see the differences related to the grant types, is that what you are worried about here? If so, I'll see about getting this updated.

I am worried about the fact that the existing tests enforce that if a client has no scopes, that we allow any scopes through, is that on purpose?

**Pull request closed without merging**